gem/optimise.py

```diff
@@ -413,3 +480,565 @@ def aggressive_unroll(expression):
+        tensors = list(mono[1])  # to be pre-evaluated
+        rest = list(mono[0])  # does not contain i
+        for t in [term for j in arg_ind_flat + (2,) for term in mono[j]]:
+            if set(t.free_indices) & quad_ind_set:
+                tensors.append(t)
+            else:
+                rest.append(t)
+        mono['pe_tensors'] = tensors = tuple(tensors)
+        mono['rest'] = tuple(rest)  # not pre-evaluated
+        # there could be duplicated pre-evaluated tensors
+        if not tensors:
+            mono['pe_result'] = tuple()
+        elif tensors in pe_results:
+            mono['pe_result'] = pe_results[tensors]
+        else:
+            # can put in a temporary tensor here
+            # mono['pe_result'] = pe_results[tensors] = (Indexed(Variable('PE'+str(len(pe_results)), (J,K)), (j,k)),)
+            # sometimes not all argument indices are present in the contraction
+            all_ind = set([fi for t1 in tensors for fi in t1.free_indices])
+            result_ind = tuple([i for i in arg_ind_flat if i in all_ind])
+            array = contract(tensors, quad_ind, result_ind)
+            pe_tensor = Indexed(Literal(array, 'PE'+str(len(pe_results))), result_ind)
+            mono['pe_result'] = pe_results[tensors] = (pe_tensor, )  # this is a tuple
+        # construct list of factor lists after pre-evaluation
+    factor_lists = []
+    for mono in monos_pe:
+        factor_lists.append(list(mono['rest'] + mono['pe_result']))
+
+    node_pe = sumproduct_2_node(factorise(factor_lists, quad_ind, arg_ind_flat))
+    node_pe = reorder(node_pe, arg_ind_flat)
+    theta_pe = count_flop(node_pe)  # flop count for pre-evaluation method
+    return (node_pe, theta_pe)
+
+
+def find_optimal_factors(monos, arg_ind_flat):
+    gem_int = OrderedDict()  # Gem node -> int
+    int_gem = OrderedDict()  # int -> Gem node
+    counter = 0
+    for mono in monos:
+        for j in arg_ind_flat:
+            # really this should just have 1 element
+            for n in mono[j]:
+                if n not in gem_int:
+                    gem_int[n] = counter
+                    int_gem[counter] = n
+                    counter += 1
+    if counter == 0:
+        return tuple()
+    # add connections (list of tuples)
+    edges_sets_list = []
+    # num_edges = dict.fromkeys(int_gem.keys(), 0)
+    for mono in monos:
+        # this double loop should be optimised further
+        edges_sets_list.append(tuple(
+            [gem_int[n] for j in arg_ind_flat for n in mono[j]]))
+
+    # product of extents of argument indices of nodes
+    extent = dict().fromkeys(int_gem.keys())
+    for key in extent:
+        arg_ind = set(int_gem[key].free_indices) & set(arg_ind_flat)
+        extent[key] = numpy.product([i.extent for i in arg_ind])
+    # set up the ILP
+    import pulp as ilp
+    prob = ilp.LpProblem('factorise', ilp.LpMinimize)
+    nodes = ilp.LpVariable.dicts('node', int_gem.keys(), 0, 1, ilp.LpBinary)
```

Comment:
FWIW, I think we should consider using `cvxopt` rather than pulp as a linear programming solver.  It also interfaces to glpk, but does so via the C api, rather than by invoking a command line program, and is therefore much faster.